### ISSUE TITLE
[Meta] Shift to Ember as a CMS

### ISSUE THREAD

**Issue by acorncom** [emberjs/guides#1251]
This is a placeholder issue for making a shift from Middleman to Ember (with a docs API backend).  It will grow over time, but these are the first thoughts on what we need to tackle:

### Phase 1
- [ ] build an initial Ember app to handle versioned apis (and multiple languages) (see #1257 for todos)
    - [ ] use pre-rendered HTML built using Middleman as part of builds
    - [ ] work through how to handle putting guides in our API system (using our build system)
- [ ] Fastboot integration (handles our SEO issues)
    - [ ] Caching (how to handle things so that pages don't need to render markdown on the server side for each page

### Phase 2
- [ ] shift away from Middleman completely (looks like we mainly need to handle highlight.js, which can be used either in Node or in-browser

### Phase 3
- [ ] offline docs using PouchDB ?

Other ideas / comments / suggestions about things needed here are more than welcome!

**Comment by rtablada** [comment]
In a similar vain, I've built todomvc.embergrep.com as an ember app that is serving markdown generated content.

Right now it is just a set of MD strings that are shoved in to an ENV variable but I'm hoping to actually make this a bit more stable and have an Ember addon that allows for a list of markdown files, publishing over to the final build, and lists of what posts are available as a static JSON feed.

**Comment by locks** [comment]
We could phase this.

* Phase 1, render HTML from `guides.github.com`
* Phase 2, render Markdown from `guides`
* etc

**Comment by acorncom** [comment]
Agreed on phasing. For phase 2, I just checked and highlight.js will run in Node, so we'll be able to do that on the Fastboot side when we get to that point.

**Comment by rondale-sc** [comment]
Had a thought this morning on this.  Would having this as an Ember app without fastboot cause it to be less visible?

**Comment by acorncom** [comment]
Yes, Fastboot is a requirement for this to be feasible. But per our docs meeting yesterday (still need to send in the minutes), Fastboot will be available for this project

**Comment by rondale-sc** [comment]
Ooh, neat!  What API system is currently the plan?

**Comment by acorncom** [comment]
Haven't figured that out yet ;-)  I've wondered about using Firebase (we already have an account), but haven't thought through it.  We'd need it to work with both Fastboot and in-browser, want to take a look at that?  I spent some time last night working up a prototype and it's working pretty nicely so far.  Will link in the PR in a minute

**Comment by acorncom** [comment]
Just posted a bunch of issues in my fork (and linked to the list from my PR).  This is going to be fun ;-)

**Comment by locks** [comment]
PouchDB would make it offline-accessible, which should be a goal IMO.

**Comment by acorncom** [comment]
@locks I think that's a far-future goal though.  Let's ship v1 before worrying about that too much :-)

**Comment by locks** [comment]
@acorncom I think it's as far-future as our API planning, due to the characteristics of PouchDB/CouchDB. CouchDB has a REST API already, PouchDB replication, should handle files/whatnot pretty well, scales well, it's a `document` oriented DB ;P
We're definitely shipping the "turbolinks" MVP first though.

**Comment by acorncom** [comment]
@locks think we should close this as well or leave for now?

**Comment by locks** [comment]
Leave it for brainstorming ;)

**Comment by mswanson** [comment]
I am currently converting our marketing site which is a Middleman generated site into an Ember app. I'm basing my work loosely on what the Ember Weekend guys have done; using mirage & fixtures for the "backend". Eventually we need to move to something like a Firebase backend and I want to get Fastboot working asap. I'm still learning parts of Ember but it would be very interested in helping out with the effort.

**Comment by locks** [comment]
Hi @mswanson. I also pinged the Ember Weekend crew to brainstorm with us in trying to find the easiest migration path. Currently, we decided to first try to tweak Middleman into the structure we want (with versions and translations.) This should make it to maintain the current guides, and any future architecture migration.
If you'd like to discuss in more detail, come by #documentation on the Slack group!
